File: internal/domain/conversation/conversation.go
package conversation

import (
	"time"

	"gorm.io/gorm"
)

// Conversation 会话实体
type Conversation struct {
	ID        string         `gorm:"primaryKey;type:varchar(36)" json:"id"`
	OrgID     string         `gorm:"type:varchar(36);index" json:"org_id"`
	Type      string         `gorm:"type:varchar(20);not null" json:"type"` // direct, group
	Name      string         `gorm:"type:varchar(100)" json:"name,omitempty"`
	IsPublic  bool           `gorm:"default:false" json:"is_public"`         // 群聊是否公开可搜索
	CreatedBy string         `gorm:"type:varchar(36);not null" json:"created_by"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`
}

func (Conversation) TableName() string {
	return "conversations"
}

// Participant 参与者
type Participant struct {
	ConversationID  string    `gorm:"primaryKey;type:varchar(36)" json:"conversation_id"`
	ParticipantType string    `gorm:"primaryKey;type:varchar(20)" json:"participant_type"` // user, agent
	ParticipantID   string    `gorm:"primaryKey;type:varchar(36)" json:"participant_id"`
	Role            string    `gorm:"type:varchar(20);default:member" json:"role"`         // owner, admin, member
	IsMuted         bool      `gorm:"default:false" json:"is_muted"`                       // 是否被禁言
	MutedUntil      *time.Time `json:"muted_until"`                                        // 禁言到期时间
	JoinedAt        time.Time `json:"joined_at"`
}

func (Participant) TableName() string {
	return "participants"
}

// ParticipantRole 参与者角色
const (
	RoleOwner  = "owner"
	RoleAdmin  = "admin"
	RoleMember = "member"
)

// JoinRequest 群聊加入请求
type JoinRequest struct {
	ID             string         `gorm:"primaryKey;type:varchar(36)" json:"id"`
	ConversationID string         `gorm:"type:varchar(36);not null;index" json:"conversation_id"`
	UserID         string         `gorm:"type:varchar(36);not null;index" json:"user_id"`
	Status         string         `gorm:"type:varchar(20);not null;default:pending" json:"status"` // pending, accepted, rejected
	Message        string         `gorm:"type:varchar(500)" json:"message"`                        // 申请消息
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
}

func (JoinRequest) TableName() string {
	return "join_requests"
}

// JoinRequestStatus 加入请求状态
const (
	JoinStatusPending  = "pending"
	JoinStatusAccepted = "accepted"
	JoinStatusRejected = "rejected"
)

// Invitation 邀请成员
type Invitation struct {
	ID             string         `gorm:"primaryKey;type:varchar(36)" json:"id"`
	ConversationID string         `gorm:"type:varchar(36);not null;index" json:"conversation_id"`
	InviterID      string         `gorm:"type:varchar(36);not null" json:"inviter_id"`      // 邀请人
	InviteeID      string         `gorm:"type:varchar(36);not null;index" json:"invitee_id"` // 被邀请人
	Status         string         `gorm:"type:varchar(20);default:pending" json:"status"`   // pending, approved, rejected
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
}

func (Invitation) TableName() string {
	return "invitations"
}

// InvitationStatus 邀请状态
const (
	InvitationStatusPending   = "pending"
	InvitationStatusApproved  = "approved"
	InvitationStatusRejected  = "rejected"
)

File: openAIMClient/Mac/openAIM/openAIM/Models/Conversation.swift
//
//  Conversation.swift
//  openAIM
//
//  Created by Claude on 2026/3/7.
//

import Foundation

/// 会话模型
struct Conversation: Codable, Identifiable, Hashable {
    let id: String
    let name: String?
    let type: ConversationType
    let organizationId: String?
    let isPublic: Bool?
    let createdBy: String?
    let createdAt: Date
    let updatedAt: Date?
    var participants: [ConversationParticipant]?
    var lastMessage: Message?
    var unreadCount: Int?
    var existing: Bool?  // 标识是否是已存在的会话（后端返回）
    var isDissolved: Bool?  // 标识群聊是否已解散（本地标记）
    // 注意：不定义 CodingKeys，让 convertFromSnakeCase 自动处理
}

/// 会话类型
enum ConversationType: String, Codable {
    case direct = "direct"
    case group = "group"
}

/// 会话参与者（API响应格式）
struct ConversationParticipant: Codable, Identifiable, Hashable {
    let id: String
    let type: String
    let name: String?
    let avatar: String?
    let role: ParticipantRole?
    let isMuted: Bool?
    let mutedUntil: Date?
}

/// 参与者角色
enum ParticipantRole: String, Codable {
    case owner = "owner"
    case admin = "admin"
    case member = "member"
}

/// 会话参与者（数据库格式）
struct Participant: Codable, Identifiable, Hashable {
    let id: String
    let conversationId: String
    let participantType: ParticipantType
    let participantId: String
    let joinedAt: Date
}

/// 参与者类型
enum ParticipantType: String, Codable {
    case user = "user"
    case agent = "agent"
}

/// 创建会话请求
struct CreateConversationRequest: Codable {
    let name: String?
    let type: ConversationType
    let orgId: String?
    let isPublic: Bool?
    let participantIds: [ParticipantIdItem]

    enum CodingKeys: String, CodingKey {
        case name
        case type
        case orgId = "orgId"
        case isPublic = "isPublic"
        case participantIds
    }
}

/// 参与者ID项
struct ParticipantIdItem: Codable {
    let type: String
    let id: String
}

/// 添加参与者请求
struct AddParticipantRequest: Codable {
    let participantType: ParticipantType
    let participantId: String
}

// MARK: - 群聊搜索结果
struct GroupSearchResult: Codable, Identifiable {
    let id: String
    let name: String?
    let participantCount: Int?
    let createdBy: String?
    let creatorName: String?
    let createdAt: Date?
}

// MARK: - 群聊加入请求
struct GroupJoinRequest: Codable, Identifiable {
    let id: String
    let conversationId: String?
    let userId: String?
    let status: GroupJoinRequestStatus?
    let message: String?
    let createdAt: Date?
    var user: User?
}

enum GroupJoinRequestStatus: String, Codable {
    case pending = "pending"
    case accepted = "accepted"
    case rejected = "rejected"
}

// MARK: - 邀请
struct GroupInvitation: Codable, Identifiable {
    let id: String
    let conversationId: String?
    let conversationName: String?
    let inviterId: String?
    let inviteeId: String?
    let status: InvitationStatus?
    let createdAt: Date?
    var inviter: User?
    var invitee: User?

    enum CodingKeys: String, CodingKey {
        case id
        case conversationId = "conversation_id"
        case conversationName = "conversation_name"
        case inviterId = "inviter_id"
        case inviteeId = "invitee_id"
        case status
        case createdAt = "created_at"
        case inviter
        case invitee
    }
}

enum InvitationStatus: String, Codable {
    case pending = "pending"
    case pendingApproval = "pending_approval"
    case approved = "approved"
    case rejected = "rejected"
}